rfcs/rfc-3-layout-callbacks.md

```diff
@@ -0,0 +1,222 @@
+
+For these reasons, it makes the most sense to put it on `WidgetProps` and let the user opt-in to them, despite it possibly being a bit annoying for library authors.
+
+#### Dispatching the Event
+
+##### Management
+
+The big question for this feature is *who's in charge of the event?* Obviously, it's an internal thing that users and integration crates won't need to manage themselves. But should the responsibility fall on `WidgetManager` or `KayakContext`? And should it go through `EventDispatcher`, a custom dispatcher, or just a simple function?
+
+While it would make sense to store this on `WidgetManager`, it's not possible due to the requirement that we pass a mutable `KayakContext` to the layout event. This means the dispatching must take place within `KayakContext`. 
+
+As for which dispatcher to use, it might be worth creating a custom `LayoutEventDispatcher` to handle the layout events. The reason for this is purely organizational. It's probably better to separate out the concerns so that `KayakContext` doesn't get too bloated.
+
+##### Logic
+
+One complication with the logic is determining which widgets to notify. To keep it simple, we can just notify all widgets that have been re-rendered. 
+
+Another additional complication is that we can't just dispatch the event over all indices in `WidgetManager::dirty_nodes` since those only contain widgets that requested a re-render (i.e., their state updated or a binding marked them dirty). Iterating over only those would skip out on all the children, which may have had their own layouts changed after rendering.
+
+This means we actually need to use `WidgetManager::dirty_render_nodes` in order to get all re-rendered nodes. So instead of draining this collection in `WidgetManager::render()`, we'll need to keep them around and drain them after calculating the layout in `KayakContext::render()`— which we'll do in `LayoutEventDispatcher::dispatch(...)`.
+
+##### `LayoutEventDispatcher`
+
+The dispatcher will look something like this:
+
+```rust
+pub(crate) struct LayoutEventDispatcher;
+
+impl LayoutEventDispatcher {
+  
+  pub fn dispatch(context: &mut KayakContext) {
+    
+    let dirty_nodes: Vec<Index> = context.widget_manager.dirty_render_nodes
+      .lock()
+      .expect("Couldn't get lock on dirty render nodes!")
+      .drain(..)
+      .collect();
+    
+    for node_index in dirty_nodes {
+      // We should be able to just get layout from WidgetManager here
+      // since the layouts will be calculated by this point
+      let mut layout_event = /* ... */
+      
+      let mut widget = context.widget_manager.take(node_index);
+      let mut context_ref = KayakContextRef::new(context, Some(node_index));
+      widget.on_layout(&mut context_ref, &mut layout_event);
+      context.widget_manager.repossess(widget);
```

Comment:
### Quick Note

I was just thinking that we might want to only fire this event when the layout actually changes, but that might be a bad idea. I think only firing on a change in layout might result in inconsistent behavior (at least from a user's perspective).

We could check and see what React Native does if we really want to be consistent with how they do things.

Reply:
Just did a quick test, and it seems React Native actually _does_ only invoke this event if there is a change in layout. Is this something we want to do? Or do we just let the user check that themselves with morphorm's `GeometryChanged` flags?

Reply:
I think it makes sense to do what react native does here. Is there a performance cost to that though? 🤔

Reply:
> Is there a performance cost to that though? 🤔

If anything it might be more performant since we don't need to dynamically call a method on a trait object. At least, we can try it out this way then bench later to see how much of a hit it causes.